collector/lib/runtime-control/Config.h

```diff
@@ -0,0 +1,54 @@
+#ifndef _RUNTIME_CONTROL_CONFIG_H_
+#define _RUNTIME_CONTROL_CONFIG_H_
+
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+#include <optional>
+
+#include <storage/runtime_filters.pb.h>
+
+#include "Hash.h"
+
+#include "../ResourceSelector.h"
+
+namespace collector::runtime_control {
+
+class Config {
+ public:
+  Config() {
+    for (int i = 0; i < storage::RuntimeFilterFeatures_ARRAYSIZE; i++) {
+      kFeatureFlags_[i] = 1UL << (i + 1);
+    }
+  }
+
+  static Config& GetOrCreate();
+
+  // returns true when initialized, false for a timeout.
+  bool WaitUntilInitialized(unsigned int timeout_ms);
+
+  void Update(const storage::RuntimeFilteringConfiguration& msg);
+  bool IsFeatureEnabled(uint64_t bitMask, storage::RuntimeFilterFeatures feature);
+  bool IsFeatureEnabled(std::string cluster, std::string ns, std::string container_id, storage::RuntimeFilterFeatures feature);
+  bool IsFeatureEnabled(std::string cluster, std::string ns, storage::RuntimeFilterFeatures feature);
+  void ConfigMessageToConfig(const storage::RuntimeFilteringConfiguration& msg);
+
+ private:
+  std::mutex mutex_;
```

Comment:
I'm very confused by this mutex, it only seems to be held on `WaitUntilInitialized` and `ConfigMessageToConfig`, the latter makes changes to `default_status_map_` and `config_by_feature_`, both of which are read by `IsFeatureEnabled` and there's no sign of the mutex being locked there, this smells of race conditions.

I think what you actually want is a `std::shared_mutex` that is `std::unique_lock`ed in methods that modify members and `std::shared_lock`ed when only reading.
https://en.cppreference.com/w/cpp/thread/shared_mutex

Reply:
The mutex was added by Olivier. I haven't thought about how it should be used much yet.